### ISSUE TITLE
Benchmark and tune FUSE on NetCDF

### ISSUE THREAD

**Issue by mrocklin** [pangeo-data/pangeo#52]
We're working a lot on Zarr right now, but we should also test the performance of gcsfuse on NetCDF data on GCS.

@jhamman has already placed a 4TB dataset in `/pangeo-data/newmann-met-ensemble-netcdf` 

The use of FUSE is an open question for a number of groups using cloud systems (cc @niallrobinson) and we should probably consider it a baseline to alternatives.

https://github.com/GoogleCloudPlatform/gcsfuse/

**Comment by mrocklin** [comment]
I was able to get gcsfuse working locally (for a while) using the following:

```
export GCSFUSE_REPO=gcsfuse-`lsb_release -c -s`
echo "deb http://packages.cloud.google.com/apt $GCSFUSE_REPO main" | sudo tee /etc/apt/sources.list.d/gcsfuse.list

sudo apt-get install gcsfuse

sudo mkdir /gcs
sudo chown -R $USER.$USER /gcs
mkdir /gcs/pangeo

gcloud auth login  # assuming you have gcloud, get credentials

gcsfuse pangeo /gcs/pangeo
```

**Comment by mrocklin** [comment]
I ran one of @jhamman 's computations on the current pangeo.pydata.org setup with both FUSE and Zarr and found that Zarr was much faster.  

```python
ds = xr.open_mfdataset('/gcs/newmann-met-ensemble-netcdf/conus_ens_00*.nc',
                       engine='netcdf4', concat_dim='ensemble', chunks={'time': 50})
# or

import gcsfs
gcsmap = gcsfs.mapping.GCSMap('pangeo-data/newman-met-ensemble')
ds = xr.open_zarr(gcsmap)
ds

da_snow = ds['pcp'].where(ds['t_mean'] < 0.).resample(time='QS-Mar').sum('time')
seasonal_snow = da_snow.isel(ensemble=slice(0, 4)).groupby('time.season').mean('time').persist()
progress(seasonal_snow)
```

Fuse took around 14 minutes.  Zarr took around 40s

cc @jhamman @martindurant

**Comment by rabernat** [comment]
Um, that's a pretty big difference!

**Comment by jhamman** [comment]
This is great. One thing to remember is that the chunk size on zarr here is about 200x that of the HDF5 chunks in the FUSE example. We should create a fair comparison to rule that out but I suspect Zarr will still be more performant.

**Comment by mrocklin** [comment]
Agreed.  I may also be off in my quick experiment

**Comment by mrocklin** [comment]
We should try this again.  It would be useful to characterize the time to pull down a single chunk and to pull down many chunks.  We should also watch the logging from `gcsfuse` to get a sense for how we're wandering through the file.  This might help us characterize the performance loss and optimizations that we can do to improve performance through caching or read-ahead.

@rabernat , I'm not sure when your interns show up, but this might be a well structured experiment.  If it works then we would want to repeat it with s3fs as well.

**Comment by rabernat** [comment]
Agreed. The interns arrive Feb. 1.

**Comment by martindurant** [comment]
The code, as written, should lend itself to being easily copied to s3fs (and azure, etc.).

**Comment by mrocklin** [comment]
Some quick notes about FUSE and NetCDF performance.  

When watching logs we found that the NetCDF/HDF library made many small 4kB reads when loading metadata, these were mostly sequential, with a few random jumps around (perhaps 3 jumps for 50 4kB reads).  Then when reading actual data it read larger chunks (64kB) in sequence.

Fortunately the `gcsfs` library already does a bit of look-ahead.  Whenever you ask for a bit of data it actually goes ahead and reads the next 5MB.  If your subsequent reads are in that 5MB then they are free.  As a result we were able to read data much more quickly than if everything were done naively.  However this still wasn't optimal, both because we jumped around a bit (currently if you jump out of the current 5MB buffer it throws it away), and because 5MB can be a long read when we do just want a small amount of data.  

In the future it would be nice to build more sophisticated buffering logic that did what it could with a fixed amount of RAM.  This might include doing a bit of look-ahead, keeping old bits of the file in case we come back to them, etc..  Ideally the logic here would be informed by looking at logs from actual reads off of NetCDF files in the wild.

Because there are several cloud providers we would not want to bake this logic into a single library like `gcsfs`, `s3fs`, or `adlfs`.  Instead, I recommend that we leverage the Python file abstraction and build logic there.  That will, I think, also make this much easier to test.

```python
with gcs.open('/bucket/path/to/file.nc') as f:
    with file_buffer(f, nbytes=100e6) as f2:
        f2.seek(...)
        data = f2.read(...)
        ...
```

**Comment by martindurant** [comment]
As an additional note (and maybe bug): you will actually get 10MB per read currently by default on gcsfs. This is because reads are constrained to 5MB boundaries by the API, the buffer wants to be at least 5MB large, and so reading even 1 byte will round you up to two blocksizes' worth.

Agree on the buffer wrapper you are suggesting, I could look into that. It seems to make sense to deal specifically with the netCDF case, and have different behaviours for 64kB versus <64kB blocks.

**Comment by martindurant** [comment]
https://github.com/martindurant/gcsfs/tree/dir_fuse is progressing further on this. Having noticed that opening a dataset with xarray might open the underlying file three times, we now cache the open file objects as well as blocks containing reads <10kB.

Note that there is no eviction, memory may grow forever. On the other hand, if each file is now only opened once, memory use is limited by the number of files you try to access, and only file areas with metadata.

**Comment by martindurant** [comment]
(now with simple LRU eviction)

**Comment by martindurant** [comment]
I am having no luck in getting generally good performance out of gcsfuse, having spent many many hours trying various things now (lots of trial commits in branch "dir_fuse" on my fork).

- For a single open_dataset (`'pangeo-data/newmann-met-ensemble-netcdf/conus_ens_001.nc'`), I can get down to ~400ms, but still it takes almost 2min to open all 101 files; some have their metadata more spread out than that first case
- `ds.t_mean.data[500:600, :, :].compute()` can have throughput >40MB/s; direct read of speeds 100MB/s is possible with gcsfs , but more metadata for the specific array is needed for access, and there’s no optimum blocksize to cover every case.
- the worst case for a single array, `x[:, 0, 0].compute()` takes ~2min, and has to scan all the 10GB of x to do it
- the big calculation on all 101 .nc files
```
da_snow = ds['pcp'].where(ds['t_mean'] < 0.).resample(time='QS-Mar').sum('time')
seasonal_snow = da_snow.isel(ensemble=slice(0, 4)).groupby('time.season').mean('time').persist()
```
  is exceptionally slow, much worse than before. It is calling open_dataset about 2000 times and 
  somehow my caching is failing despite holding on to a lot of memory.

**Comment by rabernat** [comment]
cc @kaipak

**Comment by mrocklin** [comment]
Thanks for pushing on this @martindurant .  A few questions about your setup:

-  Is there a branch that you're working on?
-  Are you working from a GCE machine, for example on pangeo.pydata.org?
-  What are some next steps that you would recommend people try if someone else were to spend some time on this?

**Comment by martindurant** [comment]
Yes, my trials are at https://github.com/martindurant/gcsfs/tree/dir_fuse (plenty of debug and profile stuff in there, so beware).

I had started on pangeo.pydata , but kept loosing sessions after a while. While that was maybe down to my local internet, I did my main timings on a long-term ec2 machine. Raw download speeds with gcsfs was still ~100MB/s, so I would expect slight performance hit, not too much.

I don't know what to try  next, which is why I stopped. There is a fundamental contention between
- each GCS operation takes >100ms, regardless of the number of bytes; using a requests.Session does not seem to alleviate this
- metadata reads are small, although often the first ~5MB of a file is read every time, and spread throughout the file, read in various orders
- data reads can be of various sizes over various parts of the file, depending on the slicing (major/minor axis)
- reads passed to FUSE are of maximum size set by the OS, e.g., 64kB, 128kB. The backend can't know the intent of the read; the last read of a large block will often be smaller, and so not be distinguishable from a metadata read (which are usually smaller).
